torchdata/datapipes/iter/transform/bucketbatcher.py

```diff
@@ -92,7 +119,109 @@ def __new__(
         if sort_key is not None:
             # In-batch shuffle each bucket seems not that useful, it seems misleading since .batch is called prior.
             if in_batch_shuffle:
-                datapipe = datapipe.batch(batch_size=bucket_num, drop_last=False).map(fn=_in_batch_shuffle_fn).unbatch()
+                datapipe = datapipe.batch(batch_size=bucket_num, drop_last=False).in_batch_shuffle().unbatch()
             else:
                 datapipe = datapipe.shuffle(buffer_size=bucket_size)
         return datapipe
+
+
+def _default_len_fn(token):
+    return len(token)
+
+
+def _token_len_fn(token, len_fn):
+    return len_fn(token), token
+
+
+def _token_filter_fn(data, *, min_len, max_len):
+    length, token = data
+    return length >= min_len and length <= max_len
+
+
+@functional_datapipe("max_token_bucketize")
+class MaxTokenBucketizerIterDataPipe(IterDataPipe[DataChunk[T_co]]):
+    r"""
+    Creates mini-batches of data from a min-heap with limited size, and the total length of samples
+    returned by ``len_fn`` within each batch will be limited by ``max_token_count``.
+    (functional name: ``max_token_bucketize``). If ``min_len`` or ``max_len`` is set, the samples with
+    length that is out of ``[min_len, max_len]`` will be filtered out.
+
+    The purpose of this DataPipe is to batch samples with similar length according to ``len_fn``.
+    Min-heap is used here to make sure the samples are sorted incrementally based on the length. And,
+    the total length of samples in each batch is guaranteed to be smaller than ``max_token_count``.
+    For an example in the audio domain, it may be batching samples with similar length. Then, given the
+    ``max_token_count``, each batch may be concatenated to a Tensor with the same size and minimum padding.
+
+    Args:
+        datapipe: Iterable DataPipe being batched
+        max_token_count: Maximum length of total length of data in each batch
+        len_fn: Function to be applied to each element to get lengths. ``len(data)`` is used by default.
+        min_len: Optional minimum length to be included into each batch
+        max_len: Optional maximum length to be included into each batch.
+        buffer_size: This restricts how many tokens are taken from prior DataPipe to bucketize
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> source_dp = IterableWrapper(['1', '22', '1', '4444', '333', '1', '22', '22', '333'])
+        >>> batch_dp = source_dp.max_token_bucketize(max_token_count=5)
+        >>> list(batch_dp)
+        [['1', '1', '1', '22'], ['22', '22'], ['333'], ['333'], ['4444']]
```

Comment:
I feel like this example is kind of confusing, since the len_fn is missing and `len(str)` is used as it. Can you change it to use `1` for all samples (like `1`, `111`, `1111`), or add indication how to define the length of the string?

Reply:
Sure, I can do that. I was using different number to represent the length of each string.